build_tools/aws-sdk-code-generator/lib/aws-sdk-code-generator/rbs/shape_dictionary/shape.rb

```diff
@@ -0,0 +1,275 @@
+        def hash
+          name.hash ^ kind.hash ^ request.hash ^ ref.hash
+        end
+
+        # for Array#uniq
+        def eql?(other)
+          return false unless other.is_a?(Shape)
+          return false unless name == other.name
+          return false unless kind == other.kind
+          return false unless request == other.request
+          return false unless ref == other.ref
+
+          true
+        end
+        alias == eql?
+
+        def fetch_body
+          dictionary.service.api.fetch("shapes").fetch(name)
+        end
+
+        def as_keyword_arguments(from:)
+          shape_body = fetch_body
+          members = shape_body["members"]
+          required = shape_body["required"]
+          type_prefix = case from
+            when :operations, :waiters
+              "Types::"
+            when :types
+              ""
+            else
+              raise from.to_s
+            end
+          params = members.sort_by.with_index do |(member_name, _), i|
+            [required&.include?(member_name) ? 'a' : 'b', member_name, i]
+          end.map do |member_name, member_body|
+            member_shape_name = member_body["shape"]
+            member_shapes = dictionary[member_shape_name]
+            case from
+            when :operations, :waiters
+              opt_prefix = required&.include?(member_name) ? "" : "?"
+              opt_suffix = ""
+              member_shape = member_shapes.find(&:input?) or raise
+            when :types
+              opt_prefix = ""
+              opt_suffix = required&.include?(member_name) ? "" : "?"
+              member_shape = member_shapes.find(&:input?) or raise
+            else
+              raise from
+            end
+            "#{opt_prefix}#{underscore(member_name)}: #{type_prefix}#{member_shape.rbs_type_name}#{opt_suffix}"
+          end
+          params.join(", ")
+        end
+
+        def rbs_name
+          if structure? && !input?
+            name
+          else
+            rbs_type_name
+          end
+        end
+
+        def rbs_print_name
+          case kind
+          when :input
+            rbs_input_name
+          when :output
+            rbs_output_name
+          when :exception
+            rbs_exception_name
+          else
+            raise
+          end
+        end
+
+        # TODO: split for input and output
+        def to_rbs_returns
+          shape_body = fetch_body
+          case type = shape_body["type"]
+          when "string"
+            if shape_body["enum"]
+              "(#{shape_body["enum"].map { "\"#{_1}\"" }.join(" | ")})"
+            else
+              "::String"
+            end
+          when "integer", "long"
+            "::Integer"
+          when "float", "double"
+            "::Float"
+          when "timestamp"
+            case shape_body["timestampFormat"]
+            when "iso8601", "rfc822", nil
+              "::Time"
+            else
+              raise [name, shape_body].inspect
```

Comment:
I think it would be reasonable to assume the interface to a timestamp shape is always ::Time regardless of its format on the wire